### PR TITLE
JoinToSubqueriesVisitor ASTColumnsMatcher fix

### LINKED ISSUE #22736: Bug in COLUMNS function in query with 3+ joins
**Describe the bug**
columns('regexp') didn't return all columns by regexp where 3+ joins in query 

**Does it reproduce on recent release?**
version 21.3.4 revision 54447

**How to reproduce**

in query we have 3 columns 'test' but returned only one:

```
select columns('test') from                                                                                                                                                               
(select 1 as id, 'a' as test) a                                                                                                                                                                    
left join (select 1 as id, 'b' as test) b on b.id = a.id                                                                                                                                               
left join (select 1 as id, 'c' as test) c on c.id = a.id;                                                                                                                                               
                                                                                                                               
Query id: 2fc3ebdf-8ec9-4684-af02-3b87bfe18705
                                                   
┌─test─┐                                           
│ c    │                                           
└──────┘                                           
                                                   
1 rows in set. Elapsed: 0.007 sec.
```

but if we change query:

```
select  columns('test'), a.test as t1, b.test as t2, c.test as t3 from 
(select 1 as id, 'a' as test) a 
left join (select 1 as id, 'b' as test) b on b.id = a.id 
left join (select 1 as id, 'c' as test) c on c.id = a.id

Query id: 5bf07e05-ec0e-4017-853d-b7b23b8a6b16

┌─--a.test─┬─--b.test─┬─test─┬─t1─┬─t2─┬─t3─┐
│ a        │ b        │ c    │ a  │ b  │ c  │
└──────────┴──────────┴──────┴────┴────┴────┘

1 rows in set. Elapsed: 0.006 sec.
```

### PR DESCRIPTION
**Pull request by kitaisreal**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix `columns` function when multiple joins in select query. Closes #22736.

### CODE REVIEW

File: src/Interpreters/JoinToSubqueryTransformVisitor.cpp [hunk 1]
```diff
@@ -73,16 +74,31 @@ class ExtractAsterisksMatcher
...
+        struct AlwaysAddColumnPredicate
+        {
+            bool operator()(const String & column_name [[maybe_unused]])
+            {
+                return true;
+            }
+        };
+
+        template <typename Predicate = AlwaysAddColumnPredicate>
+        void addTableColumns(const String & table_name, Predicate && should_add_column_predicate = Predicate())
```
Comment (KochetovNicolai) [inline src/Interpreters/JoinToSubqueryTransformVisitor.cpp:86]: Why not use `std::function<bool(const String &)> = {}`?

**Review by KochetovNicolai — APPROVED** [review]
LGTM

File: src/Interpreters/JoinToSubqueryTransformVisitor.cpp [hunk 2]
```diff
@@ -73,16 +74,31 @@ class ExtractAsterisksMatcher
             }
         }
 
-        void addTableColumns(const String & table_name)
+        struct AlwaysAddColumnPredicate
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Interpreters/JoinToSubqueryTransformVisitor.cpp]: Updated.

File: src/Interpreters/JoinToSubqueryTransformVisitor.cpp (hunk 1 above)
Comment (kitaisreal) [reply src/Interpreters/JoinToSubqueryTransformVisitor.cpp]: Updated.

File: src/Interpreters/JoinToSubqueryTransformVisitor.cpp [hunk 3]
```diff
@@ -73,16 +74,28 @@ class ExtractAsterisksMatcher
...
         }
 
-        void addTableColumns(const String & table_name)
+        using ShouldAddColumnPredicate = std::function<bool (const String&)>;
+
+        /// Add columns from table with table_name into select expression list
+        /// Use should_add_column_predicate for check if column name should be added
+        void addTableColumns(
+            const String & table_name,
+            std::function<bool (const String&)> should_add_column_predicate = [](const String &){ return true; } )
```
Comment (KochetovNicolai) [inline src/Interpreters/JoinToSubqueryTransformVisitor.cpp:83]: If we use `= {}` by default, then we can just check `if (should_add_column_predicate && should_add_column_predicate(column.name))` later

Comment (kitaisreal) [reply src/Interpreters/JoinToSubqueryTransformVisitor.cpp:83]: If we does not use {} and will use return true by default, there will be no additional checks.

**Pull request merged**